AppleSource/shell_cmds-198/renice/renice.8
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)renice.8	8.1 (Berkeley) 6/9/93
.\" $FreeBSD: src/usr.bin/renice/renice.8,v 1.16 2003/02/26 20:27:24 charnier Exp $
.\"
.Dd June 9, 1993
.Dt RENICE 8
.Os
.Sh NAME
.Nm renice
.Nd alter priority of running processes
.Sh SYNOPSIS
.Nm
.Ar priority
.Op Oo Fl p Oc Ar pid ...
.Op Oo Fl g Oc Ar pgrp ...
.Op Oo Fl u Oc Ar user ...
.Nm
.Fl n Ar increment
.Op Oo Fl p Oc Ar pid ...
.Op Oo Fl g Oc Ar pgrp ...
.Op Oo Fl u Oc Ar user ...
.Sh DESCRIPTION
The
.Nm
utility alters the
scheduling priority of one or more running processes.
The following
.Ar who
parameters are interpreted as process ID's, process group
ID's, user ID's or user names.
The
.Nm Ns 'ing
of a process group causes all processes in the process group
to have their scheduling priority altered.
The
.Nm Ns 'ing
of a user causes all processes owned by the user to have
their scheduling priority altered.
By default, the processes to be affected are specified by
their process ID's.
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl g
Force
.Ar who
parameters to be interpreted as process group ID's.
.It Fl n
Instead of changing the specified processes to the given priority,
interpret the following argument as an increment to be applied to
the current priority of each process.
.It Fl u
Force the
.Ar who
parameters to be interpreted as user names or user ID's.
.It Fl p
Reset the
.Ar who
interpretation to be (the default) process ID's.
.El
.Pp
For example,
.Pp
.Dl "renice +1 987 -u daemon root -p 32"
.Pp
would change the priority of process ID's 987 and 32, and
all processes owned by users daemon and root.
.Pp
Users other than the super-user may only alter the priority of
processes they own,
and can only monotonically increase their ``nice value''
within the range 0 to
.Dv PRIO_MAX
(20).
(This prevents overriding administrative fiats.)
The super-user
may alter the priority of any process
and set the priority to any value in the range
.Dv PRIO_MIN
(\-20)
to
.Dv PRIO_MAX .
Useful priorities are:
20 (the affected processes will run at the lowest priority),
0 (the ``base'' scheduling priority),
anything negative (lower values cause more favorable scheduling).
.Sh FILES
.Bl -tag -width /etc/passwd -compact
.It Pa /etc/passwd
to map user names to user ID's
.El
.Sh SEE ALSO
.Xr nice 1 ,
.Xr getpriority 2 ,
.Xr setpriority 2
.Sh STANDARDS
The
.Nm
utility conforms to
.St -p1003.1-2001 .
.Sh HISTORY
The
.Nm
utility appeared in
.Bx 4.0 .
.Sh BUGS
Non super-users cannot increase scheduling priorities of their own processes,
even if they were the ones that decreased the priorities in the first place.


CHANGELOG.md
# Version 5.028.1
	- Fixed Cmd+v shortcut
	- HostKey fingerprint as base64 encoding
	- Smoothing the HUD on resize or SplitView
	- Empty default user configuration bugs.
	- Zoom shortcuts changed.
	- Solarized theme fixes.
	
# Version 4.024.2
	- AirTerminals! Put a terminal on your remote AirPlay screen :)
	- Blink shortcuts. Use your external keyboard to move, create, or remove terminals. Configure the trigger too! View settings > Keyboard > Shortcuts for more info.
	- Want to use SmartKeys when an external keyboard is connected? Now you can switch them on and off from configuration.
	- View geometry of the screen from the overlay with each resize.
	- Switch cursor blinking on/off from Appearance settings.

	- Fixed TouchID issues when returning to the app.
	- Fixed cursor blink sequence so that it always starts as ON.
	- Fixed Host Port not resetting properly.
	- Fixed overlay getting covered in landscape mode.

	- Smoother experience thanks to cleanups and improvements.

# Version 3.021.2
	- iCloud Hosts sync. Synchronize hosts between devices. If a Host already has been synced, it provides conflict resolution. No critical data like passwords is saved.
	- Auto Lock. If enabled, when you lock/unlock your device, Blink will also be locked. Passcode and TouchID will be required to unlock the app.
	- Added ARMv7 support. Support for 32 bit devices like iPad 2, 3, iPhone 5, etc... We will publish depending on how well it performs!
	- Added IPv6 support for hosts.
	- Share Public Encryption Keys. You can now share the public key from the Keys section to other apps, like Mail.

	- Updated Fira Code font to v1.204.
	- Improved error checking on Themes and Font uploads. Auto correct if the GH URL is not a raw one.

	- Fixed bug with password not getting saved on host creation.
	- Fixed crash when hitting arrows with landscape keyboard on Plus devices.
	- Rolled back LC_CTYPE enforcement on server.

# Version 2.109 / 3.0
	- NEW Autorepeat for normal keys on external keyboard. Vim users rejoice!
	- NEW Default User for connections based on device.
	- NEW Comments for Public Keys.

	- Security updates to libraries.

	- Fixed critical bugs affecting ssh connections.

# Version 1.031
	- NEW On-screen keyboard with more space for modifiers, FKeys and Cursor keys. Redesigned for more space on the modifiers, and with a central scrollable area that handles more keys. Activate the Alternate keys by taping on the Alt key. And now tap on a modifier to activate it as a normal button, or make a long press to chain different combinations.
	- NEW Add your own Fonts & Themes! More info on (https://github.com/blinksh/fonts) and (https://github.com/blinksh/themes)
	- NEW Multistep authentication. Servers with google authenticator or similar will now connect without problems :)
	- NEW Fira Code font with ligatures included.

	- Added -l parameter to ssh for Hosts.
	- Improved message on how use hosts after adding them.
	- Fixed on-screen arrows.
	- Fixed F0 as F10 on external keyboard.
	- Fixed keys selection problem on settings
	- Fixed ssh-copy-id issue when accessing the host for the first time

# Version 1.019
	- Simplified build process.
	- New README and BUILD instructions

# Version 0.1020
	- "Get In Touch" and "About" sections completed.
	- iOS 10 support.
	- Added exception for iOS distribution.

	- Improved gesture support, specially on iPhone and iOS10.
	- MBProgressHUD updated with patch.
	- Test Terminal fixes for fonts and updates.
	- Disable font ligatures.
	- SSH Session termination fixes.

# Version 0.1010
	- Terminal customization! Customize your terminal with different default themes and fonts. Preview your changes within the Settings Preview section. Save your changes and restore them on each execution. Enjoy!

	- Create an id_rsa on first boot as default key.

	- Swipe and pinch conflicts
	- Copy changes to settings

# Version 0.931
	- Map Caps or Shift taps to Esc key.
	- Cursor keys (Home, End, PgUp, PgDown) with default to Cmd+arrow.
	- Function keys (F1-F10) with default to Cmd+number.
	- Change mapping of any function key to a different combination.

	- Improved placeholders for Hosts configurations.

	- Removed Alt as Esc from initial defaults.

# Version 0.927
	This version should complete the experience in relation to ssh and terminal configurations, mimicking a big part of what you can do in a normal shell:
	- Hosts Configuration. Preconfigure a host parameters, like user, port, key and commands.
	- Connect to a Host by specifying its name. Do "mosh plankton"
	- Overwrite parameters from host configuration from the shell. So carlos@plankton will override the user field on plankton.
	- Default Modifier keys settings changed: Ctrl is Ctrl, and Alt sends Esc. Everything else is undefined.
	- Modifier keys configuration. CAPS as Ctrl or ESC, no problem! Cmd as Ctrl? You have it! Configure everything to you liking.
	- Added secure passwords stored on Keychain to Host Configuration.
	- Exiting the session within the MCP closes the Space.

	- Fixed hang after "exec request accepted". Establishing connections should be smooth now.
	- Fixed adjustments on viewport after rotating the display.
	- Fixed Ctrl + Space sequences.
	- Fixed wrong/unexisting Ctrl sequences.
	- Added Esc+any character support.

	- New settings added to project. Will start to fill up in raw branch.

# Version 0.916
	This version contains many important bug fixes that should improve the experience a lot and reduce crashes to, hopefully, close to zero:
	- Unicode support to Blink shell. Thanks Yury!
	- Support for function keys + modifiers. Now you can do Ctrl/Alt/Shift + Arrows.
	- Improved experience with drag down to close. Let me know what you think :)
	- SSH with improved function keys in ncurses apps. htop, less, etc... should now work perfectly.
	- New help message with version and gestures.
	- Added asterisk, underscore and escape to SmartKeys.
	- Added HockeySDKResources.bundle to project for improved updates.


AppleSource/shell_cmds-198/renice/renice.c
/*
 * Copyright (c) 1983, 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static const char copyright[] =
"@(#) Copyright (c) 1983, 1989, 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#if 0
#ifndef lint
static char sccsid[] = "@(#)renice.c	8.1 (Berkeley) 6/9/93";
#endif /* not lint */
#endif

#include <sys/types.h>
#include <sys/time.h>
#include <sys/resource.h>

#include <err.h>
#include <errno.h>
#include <limits.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int	donice(int, int, int, int);
static int	getnum(const char *, const char *, int *);
static void	usage(void);

/*
 * Change the priority (nice) of processes
 * or groups of processes which are already
 * running.
 */
int
main(int argc, char *argv[])
{
	struct passwd *pwd;
	int errs, incr, prio, which, who;
	int delim;
	int priflag;

	errs = 0;
	incr = 0;
	which = PRIO_PROCESS;
	who = 0;
	argc--, argv++;
	if (argc < 2)
		usage();
	delim = 0;
	priflag = 0;

	/* incrementing priflag here ensures we only process
	   the single priority arg if it is the very first arg */
	for (; argc > 0; argc--, argv++, priflag++) {
	  /* once we've seen -- , don't process anymore switches */
	  if (0 == delim) {
	    /* -n must immediately be followed by the incremental
	       priority */
	    if (strcmp(*argv, "-n") == 0) {
	      incr = 1;
	      argc--, argv++;
	      if (getnum("priority", *argv, &prio))
		return (1);
	      continue;
	    }
		if (strcmp(*argv, "-g") == 0) {
			which = PRIO_PGRP;
			continue;
		}
		if (strcmp(*argv, "-u") == 0) {
			which = PRIO_USER;
			continue;
		}
		if (strcmp(*argv, "-p") == 0) {
			which = PRIO_PROCESS;
			continue;
		}
		if (strcmp(*argv, "--") == 0) {
		        delim = 1;
			continue;
		}
		if (0 == priflag) {
		/* if very first switch/arg and we've made it to
		   here, this must be the priority */
		  if (getnum("priority", *argv, &prio)) {
		  return(1);
		  }
		continue;
		}
	  }
		if (which == PRIO_USER) {
			if ((pwd = getpwnam(*argv)) != NULL)
				who = pwd->pw_uid;
			else if (getnum("uid", *argv, &who)) {
				errs++;
				continue;
			} else if (who < 0) {
				warnx("%s: bad value", *argv);
				errs++;
				continue;
			}
		} else {
			if (getnum("pid", *argv, &who)) {
				errs++;
				continue;
			}
			if (who < 0) {
				warnx("%s: bad value", *argv);
				errs++;
				continue;
			}
		}
		errs += donice(which, who, prio, incr);
	}
	exit(errs != 0);
}

static int
donice(int which, int who, int prio, int incr)
{
	int oldprio;

	errno = 0;
	oldprio = getpriority(which, who);
	if (oldprio == -1 && errno) {
		warn("%d: getpriority", who);
		return (1);
	}
	if (incr)
		prio = oldprio + prio;
	if (prio > PRIO_MAX)
		prio = PRIO_MAX;
	if (prio < PRIO_MIN)
		prio = PRIO_MIN;
	if (setpriority(which, who, prio) < 0) {
		warn("%d: setpriority", who);
		return (1);
	}
	return (0);
}

static int
getnum(const char *com, const char *str, int *val)
{
	long v;
	char *ep;

	errno = 0;
	v = strtol(str, &ep, 10);
	if (v < INT_MIN || v > INT_MAX || errno == ERANGE) {
		warnx("%s argument %s is out of range.", com, str);
		return (1);
	}
	if (ep == str || *ep != '\0' || errno != 0) {
		warnx("Bad %s argument: %s.", com, str);
		return (1);
	}

	*val = (int)v;
	return (0);
}

static void
usage()
{
	fprintf(stderr, "%s\n%s\n",
"usage: renice priority     [[-p] pid ...] [[-g] pgrp ...] [[-u] user ...]",
"       renice -n increment [[-p] pid ...] [[-g] pgrp ...] [[-u] user ...]");
	exit(1);
}
